### DISCUSSION TITLE
Widget ausblenden

### DISCUSSION THREAD

**Discussion by Chris25122** [Zefau/ioBroker.jarvis#2042] [General]
Ist es möglich ein Widget auszublenden und bei Bedarf wieder einzublenden ? Meine ist es das Kamera auszublenden wenn der Kamera Datenpunkt false ist und dieses erst wieder einzublenden wenn sich der Wert auf true ändert.

Ich habe es leider erfolglos über JSONata versucht den Layout Datenpunkt zu ändern.

![image](https://user-images.githubusercontent.com/48753907/200940464-1c0e483e-6c7b-47ee-8a69-af8e42854b50.png)

**Comment by mcuiobroker** [comment]
Wo hast du was eingetragen?
Es gibt bei einigen Bindings im JSON Probleme.
#2040 
Ich vermute, dass das dazugehört.
Stand: `v3.1.0-beta.17`
Bleibt nur abwarten.

**Reply by Chris25122** [reply]
Ich habe versucht über JSONata diesen Teil ( siehe Screenshot ) zu löschen ( 2. Element ist das Bild ) was das Widget erstmal komplett verschwinden lässt / löscht. Der Testlauf per Hand war erfolgreich. Ich bin mir nicht wirklich sicher ob das der Weg ist wie man das machen bzw. angehen sollte, hatte aber keine andere Idee.

![image](https://user-images.githubusercontent.com/48753907/200942404-6a753442-a57a-48ef-8fbc-0aeecfad866f.png)

**Reply by mcuiobroker** [reply]
Hab es noch nicht verstanden.
Was machst du genau?
Wo wird welcher Wert gesetzt, wann?

Wenn du im Json (expertenModus etwas löscht, ist es komplett weg und kann zurückkommen, indem man wieder in den Expertenmodus wechselt und es dort an die Stelle wieder einfügt.
Du möchtest es aber sehr wahrscheinlich per Schalter ein- / ausblenden?

Es gibt schon jetzt die Möglichkeit Widgets über CSS / Styles aus- / einzublenden.
https://mcuiobroker.gitbook.io/jarvis-infos/jarvis/besonderheiten-v3/styles/widgets/widget-ausblenden
Auch sollte es per Binding möglich sein. Dies ist aber zurzeit buggy.

**Reply by Chris25122** [reply]
Meine Idee ist es das untere Kamera Widget auszublenden wenn der obere Kamera Datenpunkt auf false ist und sobald dieser auf true geändert wird das untere Widget wieder einblenden.

"Es gibt schon jetzt die Möglichkeit Widgets über CSS / Styles aus- / einzublenden." Wie funktioniert das, hast du da einen Tipp ?
Das Löschen aus dem Json Expertenmodus war nur ein Test / Notlösung

**Comment by mcuiobroker** [comment]
![image](https://user-images.githubusercontent.com/84818934/200946472-4f019b5a-b19e-4bd7-8bf0-6f0d525320cb.png)

```
.jarvis-widget-fa38a8fc-79c3-4f52-9df1-d6bdee6883f6 {
	display: none !important;
}
```
Hast du die PRO-Version?

Vielleicht funktioniert es ja über das Gerät.
Wie heisst der DP / Gerät für die Kamera. und der DP/Gerät für den Schalter Kamera (an/aus).
Ist es ein displayImage-modul?

Trag mal in das Gerät zum Kamerabild (displayImage) den DP (Kameraschalter) so ein:
Aber anstatt true musst du false setzen, da du ja willst, dass das Bild nicht angezeigt wird.
![image](https://user-images.githubusercontent.com/84818934/200949429-670fd57d-6217-4c31-8c15-dcabc81c18b3.png)

![jarvis-v310-beta17-displayImage ausblenden 2](https://user-images.githubusercontent.com/84818934/200968932-74eddd68-3b49-46dc-8018-940979d21a0c.gif)

**Comment by Chris25122** [comment]
Ja Pro habe ich und es ist wie von dir vermutet ein displayImage-Modul.
Der Eintrag in Styles ist auch fast das was ich möchte, es blendet das Widget aus nur leider nicht auch den Kasten.
Gibt es eine Möglichkeit diesen mit auszublenden ?

![image](https://user-images.githubusercontent.com/48753907/201194512-a18c2075-6d58-48f4-83d7-c99f1e135741.png)

Die Zweite Möglichkeit mit dem Body Stil funktioniert leider nicht, ich habe es so eingetragen:
![image](https://user-images.githubusercontent.com/48753907/201193619-232afee3-d1c7-4407-bd52-ade8b8219d91.png)

**Reply by mcuiobroker** [reply]
>Die Zweite Möglichkeit mit dem Body Stil funktioniert leider nicht, 

Das funktioniert momentan nicht. Und das würde auch den komplette TAB ausblenden!

Für das Widget müsste es eine Option hier geben:
![image](https://user-images.githubusercontent.com/84818934/201195340-74eede53-d1c2-4925-989e-69d08e707949.png)
Kannst gerne ein Feature Request einstellen.

**Reply by Chris25122** [reply]
> > Die Zweite Möglichkeit mit dem Body Stil funktioniert leider nicht,
> 
> Das funktioniert momentan nicht. Und das würde auch den komplette TAB ausblenden!
> 
> Für das Widget müsste es eine Option hier geben: ![image](https://user-images.githubusercontent.com/84818934/201195340-74eede53-d1c2-4925-989e-69d08e707949.png) Kannst gerne ein Feature Request einstellen.

Genau das habe ich mir auch gedacht, dass es den kompletten Tab ausblenden würde / müsste.
Über das Layout wäre schonmal das Widget weg, weißt du ob es damit auch möglich ist das Frame außenrum auszublenden ?

**Reply by mcuiobroker** [reply]
Das Widget ist nicht weg, nur der Inhalt vom Widget wird ausgeblendet über die Geräte-Konfiguration (Binding), was du ja gemacht hast.
Mit v3.0.11 ging es über ein Zusatzprogramm CSSEffects:
https://mcuiobroker.gitbook.io/jarvis-infos/jarvis/system-effektprog./v3-effektprogramme/csseffects/einstellungen-beispiele/widgets-csseffects#displayoff

**Reply by Chris25122** [reply]
Dafür muss man aber extra ein Objekt mit dem CSS Inhalt anlegen wenn ich das richtig verstehe ?

**Reply by mcuiobroker** [reply]
Richtig, man muss dafür immer die CSS-Einträge anpassen.
Solange es nicht in den Optionen zum Widget eingefügt wurde. -> Feature request.

Das hat das Programm CSSEffects automatisch gemacht. Dies funktioniert aber mit v3.1.0-beta.17 nicht mehr, da einige Änderungen in der Struktur vorgenommen wurde und ich das Programm nicht angepasst habe.

**Reply by Chris25122** [reply]
Schade das wäre vermutlich wesentlich einfacher gewesen. Dann fehlt mir eigentlich nur noch der letzte Schritt um es damit ausprobieren zu können. Wie verbinde ich Objekt mit dem Widget und ist es egal an welcher Stelle im Iobroker ich das Objekt erstelle ?

**Reply by mcuiobroker** [reply]
Welches Objekt?

**Reply by Chris25122** [reply]
Achso also einfach im Jarvis unter CSS es eintragen. Ich dachte man braucht ein Iobroker Datenpunkt wo man es einträgt und dann den Datenpunkt quasi mit dem Widget bindet.

**Reply by mcuiobroker** [reply]
Das Problem ist, dass man den Eintrag im CSS mal eintragen muss, weil man es ausblenden will und andererseits rausnehmen muss, um es wieder anzeigen zu lassen.
Evtl. ist es mit Scripts möglich, muss ich mal probieren.

**Reply by Chris25122** [reply]
Wenn du Zeit und Lust dazu hast sehr gerne. Ich kenne mich leider dafür nicht gut genug aus.

**Reply by mcuiobroker** [reply]
https://mcuiobroker.gitbook.io/jarvis-infos/jarvis/besonderheiten-v3/scripts-3.1.x-pro/widgets-aus-einblenden

**Reply by Chris25122** [reply]
> https://mcuiobroker.gitbook.io/jarvis-infos/jarvis/besonderheiten-v3/scripts-3.1.x-pro/widgets-aus-einblenden

Gibts auch eine Möglichkeit die Widgets zu löschen ?
Das Ausblenden funktioniert super, allerdings bleiben die ausgeblendeten Flächen leer.
Ich hab 2 Instanzen die ich synchron halte. Die Erste Instanz beinhaltet alles, die zweite ist für das Tablet eine abgespeckte Version
bei der bestimme Widgets gelöscht werden sollen.

![image](https://user-images.githubusercontent.com/48753907/204334849-97bdef6b-71da-48cf-807b-ced0d418d629.png)

**Comment by Chris25122** [comment]
Vielen Lieben Dank funktioniert super ! :)

**Comment by mcuiobroker** [comment]
Also das Widget ist ausgeblendet, aber die anderen Widgets haben ja ihre Position ja nicht verändert. Also werden die an ihrer festgelegten Position angezeigt. Nur auf dem Handy werden die ausgeblendeten Widgets nicht als Leere Platzhalter angezeigt, da sie dort alle nacheinander angezeigt werden.

Um die "Leere" zu füllen muss man ein Programm schreiben, welches prüft welches Widget dahin passt, ob das Widget drunter passen würde usw. Also sehr aufwendig und zeitintensiv aufzusetzen. Man muss quasi laufende alle festgelegten Positionen von Widgets prüfen und evtl. anpassen.  Mir ist kein automatischer Prozess bekannt mit dem man es hinbekommen würde.
Dauert ehr Wochen oder Monate.

**Reply by Chris25122** [reply]
Stimmt das ist etwas sehr komplex. Ich hatte in Erinnerung, dass die anderen Widgets den Platz automatisch auffüllen, das ist aber anscheinend nicht der Fall, habe es eben nochmal per Hand versucht.
